### PR TITLE
Feature/chirps download

### PR DESCRIPTION
**Pull request by castledan**
Testing still missing

### CODE REVIEW

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+            is available.
+        start_month: int, default = None
+            Data will be donwloaded starting from month `start_month`.
+            If None, it is set to January.
+        end_month: int, default = None
+            Data will be donwloaded up to month `end_month`.
+            If None, it is set to the December if the year was specified,
+            otherwise to the month of the year for which most recent data
+            is available.
+        start_day: int, default = None
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:115]: I am doubting if we want to give the user this much flexibility or we just always download all days of the month that are available. As it would simplify some things. @turnerm also curious on your thoughts on this.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+    >>> from aatoolbox import create_country_config, CodAB, ChirpsMonthly
+    >>> country_config = create_country_config(iso3="bfa")
+    >>> codab = CodAB(country_config=country_config)
+    >>> codab.download()
+    >>> admin0 = codab.load(admin_level=0)
+    >>> geo_bounding_box = GeoBoundingBox.from_shape(admin0)
+    >>> chirps_monthly = ChirpsMonthly(country_config=country_config,
+                                    geo_bounding_box=geo_bounding_box)
+    >>> chirps_monthly.download(
+            start_year=2007,
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:567]: you should add 3 dots if you start a new line in the examples. Else the automatically built documentation shows them as grey

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+            clobber=clobber,
+        )
+
+    def _create_date_list(
+        self, start_year, end_year, start_month, end_month, start_day, end_day
+    ):
+
+        # Check whether the input dates are valid, assign dates when not
+        # specified and create list of tuples containing the range of
+        # dates of interest
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:335]: would make this the docstring instead of a comment

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+        except ValueError:
+            raise ValueError(
+                "(One of) the dates identified by the "
+                "given inputs are not valid."
+            )
+
+        if not start_avail_date <= start_date <= end_date <= end_avail_date:
+            raise ValueError(
+                "Check that the input dates are ordered in the "
+                f"following way: {start_avail_date} <= start date "
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:408]: I would also use the `start_date` and `end_date` vars in this string instead of only showing "start date"

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+            raise ValueError(
+                "If you specify the starting day, you also "
+                "need to specify the starting month."
+            )
+
+        if end_day is not None and end_month is None:
+            raise ValueError(
+                "If you specify the end day, you also "
+                "need to specify the end month."
+            )
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:361]: not sure if this would overcomplicate things, but there might be an easy way to make two tuples of ("starting", start_month, start_day, and start_year) and the same for the "end" pars. And then you don't have to duplicate the text?

**Review by Tinkaa — COMMENTED** [review]
This looks superpuper great Daniele! I haven't finished yet, but in case you are working on it tomorrow my first comments. 

Also, I assume the diffs in the yaml and requirements files will be solved with #117

**Comment by Tinkaa** [comment]
And also gonna add @turnerm as reviewer (but maybe I can finish the first pass if that makes it easier @turnerm)

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+
+    if Chirps.__doc__ is not None:
+        __doc__ = Chirps.__doc__ + __doc__
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, frequency="daily", **kwargs)
+
+
+@check_file_existence
+def _actual_download(filepath: Path, url: str, clobber: bool) -> Path:
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:615]: The `@check_file_existence` decorator was improved in #99 such that it can be part of the class. I.e. you can intend `_actual_download` one level and instead call `self._actual_download`

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+            end_year=2020,
+            start_month=10,
+            end_month,2
+            )
+    >>> chirps_monthly.process()
+    >>> chirps_monthly_data = chirps_daily.load()
+    """
+
+    if Chirps.__doc__ is not None:
+        __doc__ = Chirps.__doc__ + __doc__
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:577]: Oehh I like it! Only I think you need to add some info, as now the docstring of `ChirpsDaily` and `ChirpsMonthly` are the same. And somehow an empty bulletin seems to be added.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+        # three-letter name
+        month = calendar.month_abbr[int(month)]
+
+        if self._frequency == "daily":
+            url = (
+                "https://iridl.ldeo.columbia.edu/SOURCES/.UCSB/"
+                f".CHIRPS/.v2p0/.daily-improved/.global/."
+                f"{str(self._resolution).replace('.', 'p')}/.prcp/"
+                f"T/%28{day}%20{month}%20{year}%29%28{day}%20{month}%20{year}"
+                "%29RANGEEDGES/data.nc"
```
Comment (Tinkaa) [inline src/aatoolbox/datasources/chirps/chirps.py:509]: !!! You are not including the geobounding coordinates here, i.e. you are always downloading the whole globe!! 
I discovered this by running the example in the documentation you wrote, so would be good to also run this example yourself and check if it is giving the desired result.

**Review by Tinkaa — COMMENTED** [review]
Alright, I finished going through all the code. Amazing job @castledan !! 

One critical thing is to add the boundingbox coordinates to the download call. Good quick test to see if everything is working as you would expect, is to run the example you provide in the documentation yourself. 

My big design question is that the code becomes so much more complex because of allowing the `daily` and `monthly` frequency. I am wondering if we can improve this. I suspect @turnerm might have some good ideas as well. 
One radical thought I was having, is if we should even allow these different frequencies, or only support the download of `daily`. As in theory I would think `monthly` is just a sum of the `daily` data (did you check that by any chance @castledan?). So then it might almost be simpler to only have downloading of `daily` and maybe add a processing function to sum this to monthly? Curious to hear your thoughts!

File: docs/source/datasources/chirps.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+and monthly precipitation dataset, ranging from 1981 to near-present. It incorporates
+CHPclim (CHC in-house climatology), 0.05° resolution satellite imagery, and in-situ station
+data to create gridded rainfall time series for trend analysis and seasonal drought
+monitoring. More information are available here: `<https://www.nature.com/articles/sdata201566>`.
+
+Through the AA toolbox, it is possible to have access to daily and monthly data.
+
+The data is downloaded from `IRI's maproom
+<http://iridl.ldeo.columbia.edu/SOURCES/.UCSB/.CHIRPS/.v2p0>`_ as the platform allows
+for the selection of geographical areas.
```
[the remark this answers is not in the document]
Comment (castledan) [reply docs/source/datasources/chirps.rst:23]: Done

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 5 above)
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:333]: I'd prefer to keep it this way, as I think it's more clear for the user

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+            date_list = [
+                (d.split("-")[0], d.split("-")[1])
+                for d in pd.date_range(
+                    f"{start_year}-{start_month}",
+                    f"{end_year}-{end_month}",
+                    freq="MS",
+                )
+                .strftime("%Y-%m")
+                .tolist()
+            ]
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:396]: I'd prefer to keep it as it is, as I'm not sure whether it would change much in terms of length of the code.

File: docs/source/datasources/chirps.rst [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    geo_bounding_box = GeoBoundingBox.from_shape(admin0)
+
+Finally you can make use of the Chirps classes: two classes are
+available, depending on the time resolution of interest. Daily
+data are available with two spatial resolutions (0.05 and 0.25 degrees),
+whereas monthly data can be obtained only with a 0.05-degree resolution.
+
+Moreover, when downloading the data, you can choose start and end date
+of the dataset to be downloaded, by specifying start and end year, month,
+and day. All parameters are optional: if not specified, the start
```
[the remark this answers is not in the document]
Comment (castledan) [reply docs/source/datasources/chirps.rst:63]: I'd prefer to keep it as it is, as it is more complete: the user can decide to specify the year, but not the month, and so on.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+
+        if resolution not in (0.05, 0.25):
+            raise ValueError(
+                f"The given resolution is {self._resolution}, which is "
+                "not available. Has to be 0.05 or 0.25."
+            )
+
+        if self._frequency == "monthly" and self._resolution == 0.25:
+            logger.error(
+                f"{self._frequency}.capitalize() data with a resolution of "
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:76]: It capitalises the word (e.g. "monthly" -> "Monthly"), as it is at the beginning of the sentence.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+                "0.25 is not available. Automatically switching to 0.05 "
+                "resolution."
+            )
+            self._resolution = 0.05
+
+    # mypy will give error Signature of "download" incompatible with supertype
+    # "DataSource" due to the arguments not being present in
+    # `DataSource`. This is however valid so ignore mypy
+    def download(
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:85]: Yes, it does complain, hence the comment. I copied the same comment that was there for fewsnet.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+                .tolist()
+            ]
+
+        # Create a sentence containing information on the downloaded data
+        if len(date_list) == 0:
+            sentence_to_print = (
+                "No data will be downloaded. "
+                "There is no {self._frequency} available within the "
+                "chosen range of dates."
+            )
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:415]: I think it doesn't change much in terms of usability, but in terms of code structure, the way it is now is more consistent.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+        )
+
+        # Print information on data overwriting
+        clobber_dict = {True: "", False: "not "}
+
+        if len(date_list) > 0:
+            sentence_to_print += (
+                f" Data already present in the folder "
+                f"will {clobber_dict[clobber]}be overwritten."
+            )
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:149]: True, but I still feel it is a good addition, isn't it?

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 16]
```diff
@@ -0,0 +1,631 @@
...
+        if self._frequency == "daily":
+            for (year, month, day) in date_list:
+                self._download(
+                    year=year, month=month, day=day, clobber=clobber
+                )
+        else:
+            for (year, month) in date_list:
+                self._download(
+                    year=year, month=month, day=None, clobber=clobber
+                )
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:163]: Not sure what you mean exactly.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 17]
```diff
@@ -0,0 +1,631 @@
...
+                f"{self._frequency.capitalize()} "
+                "data will be downloaded, starting from "
+                f"{'-'.join(date_list[0])} to {'-'.join(date_list[-1])}."
+            )
+
+        return date_list, sentence_to_print
+
+    def _get_file_name(
+        self,
+        year: Optional[Union[int, str]],
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:427]: I'm not entirely sure. Is there a quick way to check this?

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 18]
```diff
@@ -0,0 +1,631 @@
...
+
+    def _get_raw_path(
+        self, year: Optional[str], month: Optional[str], day: Optional[str]
+    ) -> Path:
+        return self._raw_base_dir / self._get_file_name(
+            year=year, month=month, day=day
+        )
+
+    def _get_processed_path(self, raw_path: Path) -> Path:
+        return self._processed_base_dir / raw_path.parts[-1]
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:467]: I'm not sure whether this would make things easy as, in order to get the file name, you need to specify year, month and day, which you would need to get anyway from the raw_path.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 19]
```diff
@@ -0,0 +1,631 @@
...
+        day: Optional[Union[int, str]],
+    ) -> str:
+        # Set wildcard for year, to get general filename pattern
+        # used to find all files with pattern in folder
+        if year is None:
+            year = "*"
+        if month is None:
+            month = "*"
+        else:
+            if len(f"{month}") == 1:
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:438]: I believe this works only for integers, but here month is a string

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 20]
```diff
@@ -0,0 +1,631 @@
...
+            )
+        else:
+            try:
+                ds = xr.open_mfdataset(
+                    filepath_list,
+                    decode_times=False,
+                )
+                # include the names of all files that are included in the ds
+                ds.attrs["included_files"] = [f.stem for f in filepath_list]
+            except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:280]: Not sure what you mean. It can happen that some of the files are not existing (as the list filepath_list can be generated based on dates), so that is why I added a `FileNotFoundError`. I don't think there can be invalid nc files, as the user can only download valid ones (due to the check of the last date).

File: docs/source/datasources/chirps.rst [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+Background
+----------
+
+The Climate Hazards Center at UC Santa Barbara (`CHC <https://www.chc.ucsb.edu/>`_),
+in collaboration with scientists at the USGS Earth Resources Observation and Science
+(`EROS <https://www.usgs.gov/centers/eros>`_) Center, created the Climate Hazards group
+Infrared Precipitation with Stations `CHIRPS <https://www.chc.ucsb.edu/data/chirps>`_
+environmental record.
+
+CHIRPS is a quasi-global (50&deg;S-50°N), high resolution (0.05°), daily, pentadal,
```
[the remark this answers is not in the document]
Comment (Tinkaa) [reply docs/source/datasources/chirps.rst]: Larger than this comment, but in that case I think it would be good to understand if the pentaal and monthy data are anything else than just the sum of the daily data over the respective time period.

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Tinkaa) [reply src/aatoolbox/datasources/chirps/chirps.py:77]: Idea makes sense but then it should be within the `{}` right?

File: tests/datasources/test_chirps.py [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+        chirps = mock_chirps(frequency=frequency)
+        chirps.download(
+            start_year=start_year,
+            end_year=end_year,
+            start_month=start_month,
+            end_month=end_month,
+            start_day=start_day,
+            end_day=end_day,
+        )
+        _, kwargs_download = download_mock.call_args
```
Comment (Tinkaa) [inline tests/datasources/test_chirps.py:81]: Okay sorry, I dont have enough time atm to look at this properly but I think `call_args` returns the last call only? would it instead make more sense to return `call_args_list` as we would generally expect `download_mock` to be called several times?

Comment (Tinkaa) [reply tests/datasources/test_chirps.py:81]: then you could later on test that not only the last date is returned but all that you expect

File: tests/datasources/test_chirps.py [hunk 23]
```diff
@@ -0,0 +1,329 @@
...
+        == chirps._create_date_list(
+            start_year=_START_YEAR,
+            end_year=_END_YEAR,
+            start_month=_START_MONTH,
+            end_month=_END_MONTH,
+            start_day=_START_DAY,
+            end_day=_END_DAY,
+        )[0]
+    )
+
```
Comment (Tinkaa) [inline tests/datasources/test_chirps.py:190]: I know this becomes a lot of testing but you might also want to test when some of the input args to `_create_date_list` are None

**Review by Tinkaa — COMMENTED** [review]
Looking good! And nice tests. Such a fast learner :) 

I havent had the time to go through the testing super properly, so happy to have another look when I am back. But also happy if Monica finishes her review to merge already. 

Few small general points: 
- Since the code has so many if/else cases and "opportunities" to raise errors, you are still missing quite a bit of the code in your testing. I would add a few extra tests for these cases. With the link in Slack, you can easily see which lines are now not being tested. 
- You should merge in develop to this branch. Hopefully that will resolve the diferences on the requirement files
- The tests for python 3.10 are currently failing. So check why

File: docs/source/datasources/chirps.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (castledan) [reply docs/source/datasources/chirps.rst]: I did a couple of checks and it seems that monthly data are indeed the sum of daily data

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py]: The problem is that it is much easier to deal with dates in the string form than date object, when it comes to dates with only specified year and month: the reason is that, when converting dates between timestamps and datetime object, the day gets assigned in a weird way and it is difficult to build efficient tests.

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:77]: I checked and it works also this way

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py]: Good point. Anyway, I deleted this part, as the check happens already earlier, so no empty lists can be created.

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:154]: What do you think, @turnerm?

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:493]: I'm not sure I understood what you mean. @turnerm, any idea about Tinka's comment?

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py]: About the first point: I think that's exactly what the check with `FileNotFoundError` does.

About the second point: I don't think this can happen at the loading state, as the downloaded datasets have already been opened once. However, I will add this check to the processing method.

**Comment by castledan** [comment]
About the difference between daily and monthly data: I did some checks and it seems that monthly data are constituted just by the sum of daily data. However, I would keep both classes there, as downloading daily data requires a lot of time, and many users might be interested only in monthly data.

File: tests/datasources/test_chirps.py (hunk 22 above)
Comment (castledan) [reply tests/datasources/test_chirps.py:108]: Not sure if this is necessary, as there are already tests on `_create_date_list`. @turnerm what do you think?

**Comment by caldwellst** [comment]
Great stuff! I'm not reviewing like Monica or Tinka, but just wanted to add here that I think the way this deals with dates (start year, start month, etc.) versus the way I deal with it in NDVI or others deal with it elsewhere says to me we should have some overarching design decisions that guide how we implement individual data sources. I would hope the end API for different data sets generally resemble others so it's easier for the user to switch between and just pass the same parameters across data sets for instance.

File: tests/datasources/test_chirps.py [hunk 24]
```diff
@@ -0,0 +1,487 @@
...
+import cftime
+import numpy as np
+import pytest
+import xarray as xr
+from xarray.coding.cftimeindex import CFTimeIndex
+
+from aatoolbox import ChirpsDaily, ChirpsMonthly, GeoBoundingBox
+
+DATASOURCE_BASE_DIR = "chirps"
```
Comment (turnerm) [inline tests/datasources/test_chirps.py:13]: Either make all the globals private or public. For tests I am usually lazy and make everything public since it doesn't really matter.

File: tests/datasources/test_chirps.py [hunk 25]
```diff
@@ -0,0 +1,487 @@
...
+        )
+        _, kwargs_download = download_mock.call_args
+        url = kwargs_download["url"]
+        filepath = kwargs_download["filepath"]
+        return url, filepath
+
+    return _mock_download
+
+
+def test_download_monthly(
```
Comment (turnerm) [inline tests/datasources/test_chirps.py:111]: In this test and the next few, I think you still need to add a mock out `_get_last_available_date`

File: tests/datasources/test_chirps.py [hunk 26]
```diff
@@ -0,0 +1,487 @@
...
+        (
+            "aatoolbox.datasources.chirps.chirps."
+            "_Chirps._get_last_available_date"
+        ),
+        return_value=date(year=2025, month=5, day=5),
+    )
+    chirps = mock_chirps(frequency="daily")
+
+    with pytest.raises(ValueError):
+        chirps._check_dates_validity(
```
Comment (turnerm) [inline tests/datasources/test_chirps.py:224]: Ideally you should be testing through public methods rather than private ones

File: tests/datasources/test_chirps.py (hunk 22 above)
Comment (turnerm) [reply tests/datasources/test_chirps.py:103]: I think Tinka is correct, her suggestion is really the "right" way to do it. Plus tests should generally be on public methods.

**Review by turnerm — COMMENTED** [review]
Really nice job Daniele! I think the two main changes from the comments are:
1. Changing the date / time input (let's see what comes out of #128)
2. Exploiting the class structure a bit more -- i.e. avoiding `if frequency ==...` statements, and instead building those differences into the child classes

File: tests/datasources/test_chirps.py [hunk 27]
```diff
@@ -0,0 +1,360 @@
...
+            "r0.05_Np6Sp3Ep2Wm3.nc"
+        )
+        for MONTH in range(int(START_MONTH), int(END_MONTH) + 1)
+    ]
+
+    assert url_list == url_list_control
+    assert filepath_list == filepath_list_control
+
+
+def test_download_daily(mock_aa_data_dir, mock_country_config, mock_download):
```
Comment (turnerm) [inline tests/datasources/test_chirps.py:160]: This test needs more mocking (it still needs an internet connection), as well as `test_chirps_load`

File: tests/datasources/test_chirps.py [hunk 28]
```diff
@@ -0,0 +1,360 @@
...
+            f"{DATASOURCE_BASE_DIR}/"
+            f"abc_chirps_daily_{END_YEAR}_{END_MONTH}_{END_DAY}_"
+            "r0.05_Np6Sp3Ep2Wm3.nc"
+        )
+    ]
+
+    mocker.patch(
+        (
+            "aatoolbox.datasources.chirps.chirps."
+            "_Chirps._get_to_be_loaded_path_list"
```
Comment (turnerm) [inline tests/datasources/test_chirps.py:346]: I wouldn't mock this, as I think that's the main thing to test here

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 29]
```diff
@@ -0,0 +1,572 @@
...
+            f"{location_url}"
+            f"T/%28{day}%20{month_name}%20{year}%29%28{day}"
+            f"%20{month_name}%20{year}"
+            "%29RANGEEDGES/data.nc"
+        )
+
+        return url
+
+    @check_file_existence
+    def _process(self, filepath: Path, ds, clobber: bool) -> Path:
```
Comment (turnerm) [inline src/aatoolbox/datasources/chirps/chirps.py:562]: This method isn't covered by the tests, maybe you need to implement a test for both monthly and daily?

File: tests/datasources/test_chirps.py [hunk 30]
```diff
@@ -0,0 +1,360 @@
+"""Tests for the Chirps module."""
```
Comment (turnerm) [inline tests/datasources/test_chirps.py:1]: The coverage overall is very good. The only thing really missing is testing a few edge cases where you raise an error. But you can skip it if you want.

**Review by turnerm — COMMENTED** [review]
Yay this looks really great Daniel, I also I think you shaved off like 100 lines!

Sorry just one more round of nitpicky comments from me. A couple to add to the list:
- some methods are missing return types, those should be added in (also to any `abstractmethods`)
- make sure you update the changelog

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 31]
```diff
@@ -0,0 +1,572 @@
+"""Class to download and load CHIRPS observational precipitation data."""
+import calendar
+import logging
+from abc import abstractmethod
+from datetime import date, datetime
+from pathlib import Path
+from typing import Optional
+
+import cftime
```
Comment (turnerm) [inline src/aatoolbox/datasources/chirps/chirps.py:9]: Oops should have noticed this earlier, but cftime needs to be added to the list of dependencies in `setup.py`, and the requirements files updated.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 32]
```diff
@@ -0,0 +1,572 @@
...
+    ----------
+    country_config : CountryConfig
+        Country configuration
+    geo_bounding_box: GeoBoundingBox
+        the bounding coordinates of the area that should be included in the
+        data.
+
+    Examples
+    --------
+    >>> from aatoolbox import create_country_config, CodAB, ChirpsMonthly
```
Comment (turnerm) [inline src/aatoolbox/datasources/chirps/chirps.py:375]: Add GeoBoundingBox

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 33]
```diff
@@ -0,0 +1,572 @@
...
+        pass
+
+    def _get_last_available_date(self):
+        pass
+
+    def _get_downloaded_path_list(self):
+        """Create a list with all raw data downloaded."""
+        # Get the path where raw data is stored
+        raw_path = self._get_raw_path(year=None, month=None, day=None)
+        return list(raw_path.parents[0].glob(raw_path.name))
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:326]: I think it'd be handy to keep it as it is for the method `process()`. This step requires very little time to run, so, instead of having the user needing to specify `start_date` end `end_date` for the method, it may be just easier to process all downloaded data. What do you think?

In the `load()` method I agree with you, I indeed changed it.

**Comment by turnerm** [comment]
Regarding the 3.10 error: I think it's related to [this issue](https://stackoverflow.com/questions/71006708/getting-sslv3-alert-handshake-failure-when-trying-to-connect-to-imap), which means:

1. When the date check is mocked the tests will be fixed, however...
2. The problem still needs to be solved for Python 3.10. A workaround could be:

```python
import ssl
from urllib.request import urlopen

context=ssl.create_default_context()
context.set_ciphers("DEFAULT")
result = urlopen(url, context=context)
df = pd.read_csv(result)
```

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 34]
```diff
@@ -0,0 +1,572 @@
...
+            start_date=start_date,
+            end_date=end_date,
+        )
+
+        date_list = pd.date_range(
+            start_date, end_date, freq=frequency_dict[self._frequency]
+        ).tolist()
+
+        # Create a sentence containing information on the downloaded data
+        sentence_to_print = (
```
[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:239]: Ah I see. I would still try to avoid the message passing: it's kind of confusing when methods return two different things, also it means that `_get_to_be_loaded_path_list` doesn't actually use the message. 

What you could do instead is either have a boolean flag in `_create_date_list` to print the message or not. Even better, pass the logging level that you want to use to print the message. You could make the default debug, and `download` would pass `info`.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 35]
```diff
@@ -0,0 +1,572 @@
...
+            end_date=end_date,
+        )
+
+        logger.info(sentence_to_print)
+
+        # Data download
+        for d in date_list:
+            last_filepath = self._download_prep(d=d, clobber=clobber)
+
+        return last_filepath.parents[0]
```
[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:113]: Ah that's true. I'm wondering, maybe you can pull `_check_dates_validity` out of `_create_date_list` and have it run before? This way nesting is reduced a bit and it will also be more obvious that indeed the dates were checked.

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:326]: I was trying to think of what we do in other cases, and I realized that for NDVI the start and end dates are [passed to the constructor](https://github.com/OCHA-DAP/pa-aa-toolbox/pull/100/files#diff-f240ac32a46252d1b17ad5c1779b358f314b935636d9629660270b30b86384e7R77). (So sorry, I should have checked that earlier). What do you think of doing that instead?

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 36]
```diff
@@ -0,0 +1,572 @@
...
+
+        return ds.rio.write_crs("EPSG:4326", inplace=True)
+
+    def _download_prep(self, d, clobber):
+        # Preparatory steps for the actual download
+        year = str(d.year)
+        month = f"{d.month:02d}"
+        day = f"{d.day:02d}"
+        output_filepath = self._get_raw_path(year=year, month=month, day=day)
+        output_filepath.parent.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:218]: You would just need the directory, which in principle should be simpler. But if it's too complicated then you can leave it.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 37]
```diff
@@ -0,0 +1,709 @@
...
+    ... end_month,2
+    ... )
+    >>> chirps_monthly.process()
+    >>> chirps_monthly_data = chirps_daily.load()
+    """
+
+    if _Chirps.__doc__ is not None:
+        __doc__ = _Chirps.__doc__ + __doc__
+
+    def __init__(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:394]: I'm not sure about what you exactly mean

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 38]
```diff
@@ -0,0 +1,572 @@
...
+import xarray as xr
+
+from aatoolbox.config.countryconfig import CountryConfig
+from aatoolbox.datasources.datasource import DataSource
+from aatoolbox.utils.check_file_existence import check_file_existence
+from aatoolbox.utils.geoboundingbox import GeoBoundingBox
+
+logger = logging.getLogger(__name__)
+
+frequency_dict = {"daily": "D", "monthly": "SMS"}
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:21]: Do you mean that this dictionary should not be defined as a constant, but rather as a variable in both classes?

[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:21]: Yes exactly!

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:239]: Sorry, I don't understand the last part.

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 31 above)
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:9]: Ah so sorry, I meant setup.cfg

File: src/aatoolbox/datasources/chirps/chirps.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:239]: No worries, it was not super clear. What you could do is make one of the parameters to `_create_date_list` be `logging_level`, and then you can call `logging.log(logging_level, msg)`.

**Comment by Tinkaa** [comment]
@castledan sorry I lost the status of this, is it ready to be merged?

**Comment by castledan** [comment]
I think so, unless you or @turnerm have other comments.

**Comment by turnerm** [comment]
Oops so sorry that I've dropped this forever, I will take a look!

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 39]
```diff
@@ -0,0 +1,646 @@
...
+                "indicate the range for which CHIRPS data are "
+                "currently available."
+            )
+
+    def _get_file_name_base(
+        self,
+        year: Optional[str],
+        month: Optional[str],
+    ) -> str:
+        # Set wildcard for year, to get general filename pattern
```
Comment (turnerm) [inline src/aatoolbox/datasources/chirps/chirps.py:261]: Aren't the wildcards not required anymore?

**Review by turnerm — APPROVED** [review]
Hey Daniele this looks really awesome, I just left a couple of small comments. So sorry again for the the delay from my side, I took way too long! 

I think it's definitely ready to go, the only thing I would say for next time is to avoid too much nesting, i.e. methods that call other methods, that call other methods, etc. It makes it hard to follow the flow and debug. It's usually better to try and stay only one or two methods deep, return the values you need, and pass them to the next method. Hope that makes sense.

File: src/aatoolbox/datasources/chirps/chirps.py [hunk 40]
```diff
@@ -0,0 +1,646 @@
...
+    ):
+        super().__init__(
+            country_config=country_config,
+            geo_bounding_box=geo_bounding_box,
+            frequency="monthly",
+            resolution=0.05,
+            start_date=start_date,
+            end_date=end_date,
+        )
+        self._date_range_freq = "SMS"
```
[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:441]: Do you mean that it should be passed by the user? I thought it is better to keep it as a private variable, as the user is not supposed to change it. Also, the value is different between ChirpsDaily and ChirpsMonthly: how can I set it in the base class?

[the remark this answers is not in the document]
Comment (turnerm) [reply src/aatoolbox/datasources/chirps/chirps.py:441]: No sorry my bad I wasn't super clear, what I meant is it should be a required parameter of the base class (_Chirps) init, so you have to pass it from the child classes, and then in the base class set self._date_range_freq

[the remark this answers is not in the document]
Comment (castledan) [reply src/aatoolbox/datasources/chirps/chirps.py:441]: Understood!

**Comment by Tinkaa** [comment]
woohoo great job,exicted to use it!! 
One note, that you can now also delete the branch after you have merged it

**Comment by castledan** [comment]
Was not sure about it, then I'll delete it!
